src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.JsonTypeInfoOfT.cs

```diff
@@ -0,0 +1,47 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if !NETCOREAPP
+using System.Diagnostics;
+#endif
+using System.IO;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public sealed partial class JsonContent
+    {
+        private sealed class JsonContentSerializer<TValue> : JsonContentSerializer
```

Comment:
The `JsonContentSerializer` type appears redundant and costs one additional allocation per content instance. Is it not possible to just put it all inside the one `JsonContent` class?

Reply:
I looked into that, but it was problematic. One approach is to inherit from JsonContent, which was discussed on the original issue. However, that requires unsealing JsonContent, which wasn't an approved API surface change. It also has problems with making non-virtual properties virtual (not allowed due to `call` vs `callvirt` compatibility) or needing to do some weird indirection on the getters.

The other approach would be multiple fields in `JsonContent` that cover all forms, combined with a lot of `if` statements to use the right form. This seemed messy and counter to the idea of polymorphism, which is designed to reduce such mess.

This approach seemed cleaner at the cost of a very lightweight additional abstraction layer. That said, I'm happy to go with one of the other two ways if it is preferred.

Reply:
I've updated to use an inheritence pattern from `JsonContent` with two internal subclasses, `JsonContentUntyped` and `JsonContent<T>`. It actually wasn't as messy as I feared. The only downside is `JsonContent` is now `abstract`, but the constructors are all `private protected` so consumers still cannot inherit from it.